pubsublite/internal/telemetry/publisher.go

```diff
@@ -19,54 +19,132 @@ package telemetry
 
 import (
 	"context"
+	"fmt"
 
 	"cloud.google.com/go/pubsub"
 	"go.opentelemetry.io/otel"
 	"go.opentelemetry.io/otel/attribute"
 	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/metric"
 	"go.opentelemetry.io/otel/propagation"
 	semconv "go.opentelemetry.io/otel/semconv/v1.18.0"
 	"go.opentelemetry.io/otel/trace"
+
+	"github.com/elastic/apm-queue/pubsublite/internal/pubsubabs"
 )
 
-type producerHandler = func(context.Context, *pubsub.Message) *pubsub.PublishResult
+const (
+	producedKey = "producer.messages.produced"
+	erroredKey  = "producer.messages.errored"
+
+	instrumentName = "github.com/elastic/apm-queue/pubsublite"
+)
 
-// Publisher adds telemetry data to messages published
-func Publisher(ctx context.Context, tracer trace.Tracer, msg *pubsub.Message, h producerHandler, attrs []attribute.KeyValue) *pubsub.PublishResult {
+// ProducerMetrics hold the metrics that are recorded for producers
+type ProducerMetrics struct {
```

Comment:
nit: the package name is `publisher`, going by that flow we can have it named `PublisherMetrics` (or vise versa)? (feel free to ignore)